DOCS-#3778: Improve documentation for Modin vs. Dask (#3779)
modin-project/modin: docs/getting_started/modin_vs_dask_vs_koalas.rst

**Modin supports all of the above pandas API functions, as well as others, with** `more than 90% <https://github.com/modin-project/modin#pandas-api-coverage>`_ **coverage of the pandas API.**  Modin additionally acts as a drop-in replacement for pandas, such that even if the API is not yet supported, it still works by falling back to running vanilla pandas. One of the key features of being a drop-in replacement is that not only will it work for existing code, if a user wishes to go back to running pandas directly, they are not locked in to using Modin and can switch between Modin and pandas at no cost. In other words, scripts and notebooks written in Modin can be converted to and from pandas as the user desires by simply replacing the import statement.

> Computation Semantics

Reviewer: ```suggestion
Execution Semantics
```

> Modin vs. Dask DataFrame and Modin vs. Koalas

Author: Possible title: "How Modin differs from Dask DataFrame and Koalas?"

> Modin vs. Dask DataFrame and Modin vs. Koalas

Reviewer 2: Originally it was Modin vs. Dask Dataframe vs. Koalas. I expanded it out for SEO reasons. We can change it back.

> Modin vs. Dask DataFrame and Modin vs. Koalas

Author: ```suggestion
Modin vs. Dask DataFrame vs. Koalas
```

> Modin vs. Dask DataFrame and Modin vs. Koalas

Author: Sure let's change it back then.

===============================================

> Libraries such as Dask DataFrame (DaskDF for short) and Koalas aim to support the pandas API on top of distributed computing frameworks, Dask and Spark respectively. Instead, Modin aims to preserve the pandas API and semantics as is, while abstracting away the details of the distributed computing framework underneath. Thus, the aims of these libraries are fundamentally different.

Author: ```suggestion
Libraries such as `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF for short) and `Koalas <https://koalas.readthedocs.io/en/latest/>`_ aim to support the pandas API on top of distributed computing frameworks, Dask and Spark respectively. Instead, Modin aims to preserve the pandas API and semantics as is, while abstracting away the details of the distributed computing framework underneath. Thus, the aims of these libraries are fundamentally different.
```

Author: Added link at first mention

Reviewer 2: LGTM

Specifically, Modin enables pandas-like

> * row and column-parallel operations, unlike DaskDF and Koalas that only support row-parallel operations
> * indexing & ordering semantics, unlike DaskDF and Koalas that deviate from these semantics
> * eager execution, unlike DaskDF and Koalas that provide lazy execution

Author: ```suggestion
* row and column-parallel operations, unlike DaskDF and Koalas which only support row-parallel operations
* indexing & ordering semantics, unlike DaskDF and Koalas which deviate from these semantics
* eager execution, unlike DaskDF and Koalas which provide lazy execution
```

Author: Unclear what `indexing & ordering semantics` means without prior context, can we say `indexing & ordering capabilities` instead?

Reviewer 2: If we are changing that -> which, we should also add a comma before each which.

Re: semantics vs. capabilities, it is not really about capabilities but it is about what is the expected behavior (hence semantics). We can change it to "behavior" if semantics is hard to understand.

Author: Added comma before which.

Author: I agree "behavior" is a better way of framing this, semantics can mean many things.

As a result, Modin's coverage is `more than 90% <https://github.com/modin-project/modin#pandas-api-coverage>`_ of the pandas API, while DaskDF and Koalas' coverage is about 55%. 

.. figure:: ../img/api_coverage_comparison.svg
   :align: center
   :alt: Percentage coverage of the pandas API after deduplication

> For more technical details please see our VLDB 2022 paper, referenced `here <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_. 

Author: ```suggestion
For more technical details please see our VLDB 2022 research paper, referenced `here <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_. 
```

Reviewer 2: LGTM

Brief Overview of DaskDF and Koalas
-----------------------------------

> Dask's `DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF) is effectively a meta-frame, partitioning and scheduling many smaller ``pandas.DataFrame`` objects. Users construct a task graph of dataframe computation step by step and then trigger computation using the ``compute`` function.

Author: Unclear if the term `meta-frame` makes sense to readers.

Reviewer 2: How about "meta DataFrame"? I don't think I changed this from the previous text.

Author: ```suggestion
Dask's `DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF) is effectively a meta-DataFrame, partitioning and scheduling many smaller ``pandas.DataFrame`` objects. Users construct a task graph of dataframe computation step by step and then trigger computation using the ``compute`` function.
```

> Dask's `DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF) is effectively a meta-frame, partitioning and scheduling many smaller ``pandas.DataFrame`` objects. Users construct a task graph of dataframe computation step by step and then trigger computation using the ``compute`` function.

Author: ```suggestion
Dask's `DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF) acts as a master dataframe, partitioning and scheduling many smaller ``pandas.DataFrame`` objects. Users construct a task graph of dataframe computation step by step and then trigger computation using the ``compute`` function.
```

Reviewer 2: I would prefer "meta DataFrame". We should avoid the use of the term "master" due to racist connotations.

**Modin supports both row, column, and block-oriented partitioning and parallelism**. That is, the dataframe can be conceptually broken down as groups of rows, groups of columns, or both groups of rows and groups of columns (effectively a block or sub-matrix). Modin will transparently reshape the partitioning as necessary for the corresponding operation, based on whether the operation is row-parallel, column-parallel, or cell-parallel (independently applied to each unit cell). This allows Modin to support more of the pandas API and do so efficiently. Due to the finer-grained control over the partitioning, Modin can support a number of operations that are very challenging to parallelize in row-oriented systems (e.g., ``transpose``, ``median``, ``quantile``). This flexibility in partitioning also gives Modin tremendous power to implement efficient straggler mitigation and improve utilization over the entire cluster.

> API Coverage

Author: Wondering if we should have the "API Coverage" section before the "Partitioning and Parallelization" section. 
Right now it feels like the partitioning is not well motivated, users don't care about partitioning for its own sake but for the capabilities, they afford (note that this page also sits in the "Getting Started" guide). If we put API first, we can say here is the higher API coverage, and how do we do this through partitioning. Perhaps it is worth adding a sentence at the top of the "Partitioning and Parallelization" section that says something like *"Modin, DaskDF, Koalas are all examples of parallel dataframe system. Parallelism is achieved through partitioning a large dataframe into smaller units that can be operated on. As a result, the partitioning scheme chosen by the library dictates what pandas functions can or can not be supported."*

Reviewer 2: It is hard to understand why the API coverage is higher without having described row/column/cell-based partitioning first, so I am more in favor of having that be first. 

That said, I like your idea of adding a sentence at the start of partitioning for why we're talking about it, but small rewording/fixing typos: _Modin, DaskDF, Koalas are all examples of parallel dataframe systems. Parallelism is achieved by partitioning a large dataframe into smaller ones that can be operated on in parallel. As a result, the partitioning scheme chosen by the system dictates the pandas functions that can or can not be supported.__

Author: Added the header paragraph to Partitioning section

------------

> 

Author: Possibly adding a header sentence saying something like *"One of the great things about pandas is the wide array of convenient functionalities, with more than 600+ APIs for XYZ. However, it is not trivial to implement all of these APIs in a scalable dataframe system."*

Reviewer 2: Yes, rewording/expanding: "_One of the key benefits of pandas is its versatility, due to the wide array of operations, with more than 600+ API operations for data cleaning, feature engineering, data transformation, data summarization, data exploration, and machine learning. However, it is not trivial to develop scalable implementations of each of these operations in a dataframe system._"

Author: Looks great! Adding this in: 

```suggestion

One of the key benefits of pandas is its versatility, due to the wide array of operations, with more than 600+ API operations for data cleaning, feature engineering, data transformation, data summarization, data exploration, and machine learning. However, it is not trivial to develop scalable implementations of each of these operations in a dataframe system.
```

> **DaskDF and Koalas only implements about** `55%  <https://arxiv.org/abs/2001.00888>`_ **of the pandas API**; they do not implement certain APIs that would deviate from the row-wise partitioning approach, or would be inefficient with the row-wise parallelization. For example, Dask does not implement ``iloc``, ``MultiIndex``, ``apply(axis=0)``, ``quantile``(only approximate quantile is available), ``median``, and more. Given DaskDF's row-oriented architecture, ``iloc``, for example, can technically be implemented, but it would be inefficient, and column-wise operations such as ``apply(axis=0)`` would be impossible to implement. Similarly, Koalas does not implement ``apply(axis=0)`` (it only applies the function per row partition, giving a different result), ``quantile``,``median`` (only approximate quantile/median is available), ``MultiIndex``, ``combine``, ``compare`` and more.

Author: ```suggestion
**DaskDF and Koalas only implements about** `55%  <https://arxiv.org/abs/2001.00888>`_ **of the pandas API**; they do not implement certain APIs that would deviate from the row-wise partitioning approach, or would be inefficient with the row-wise parallelization. For example, Dask does not implement ``iloc``, ``MultiIndex``, ``apply(axis=0)``, ``quantile`` (only approximate quantile is available), ``median``, and more. Given DaskDF's row-oriented architecture, ``iloc``, for example, can technically be implemented, but it would be inefficient, and column-wise operations such as ``apply(axis=0)`` would be impossible to implement. Similarly, Koalas does not implement ``apply(axis=0)`` (it only applies the function per row partition, giving a different result), ``quantile``, ``median`` (only approximate quantile/median is available), ``MultiIndex``, ``combine``, ``compare`` and more.
```

Reviewer 2: LGTM

**DaskDF additionally does not support multi-indexing or sorting.** 
DaskDF sorts the data based on a single set of row labels for fast row lookups, and builds an indexing structure based on these row labels. Data is both logically and physically stored in the same order. As a result, DaskDF does not support a `sort` function.

**Modin reproduces the intuitive behavior in pandas where the order of the DataFrame is preserved, and supports multi-indexing.** Enforcing ordering on a parallel dataframe system like Modin requires non-trivial effort that involves decoupling of the logical and physical representation of the data, enabling the order to be lazily kept up-to-date, but eagerly computed based on user needs (See Section 4.2 in `our recent paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_). Modin abstracts away the physical representation of the data and provides an ordered view that is consistent with user's expectations.

> Compatibility with Computational Frameworks

Author: ```suggestion
Compatibility across Computational Frameworks
```

Reviewer 2: This is a nit but I would keep the *with* -- because Modin works with (i.e., on top of) multiple computational frameworks.

Author: makes sense. let's keep the *with*.

**Modin's highly modular design is architected to run on a variety of systems, and support a variety of APIs.** The goal for the extensible design is that users can take the same notebook or script and seamlessly move between different clusters and environments, with Modin being able to support the pandas API on your preexisting infrastructure. Currently, Modin support running on Dask's compute engine in addition to Ray. The modular design makes it easier for developers to different execution engines or compile to different memory formats. Modin can run on a Dask cluster in the same way that DaskDF can, but they differ in the ways described above. In addition, Modin is continually expanding to support popular data processing APIs (SQL in addition to pandas, among other DSLs for data processing.) while leveraging the same underlying execution framework. Modin's flexible architecture also means that as the `pandas API continues to evolve <https://data-apis.org/blog/announcing_the_consortium/>`_, Modin can quickly move towards supporting new versions of the pandas API.

> . figure:: ../img/performance-all-supported.svg

Author: ```suggestion
.. figure:: ../img/performance-all-supported.svg
```

Reviewer 2: LGTM

   :align: center
   :alt: Scalability of operators supported by Modin and other systems

Performance Comparison
----------------------

> **On operations supported by all systems, Modin provides substantial speedups.** Thanks to its optimized design, Modin is able to take advantage of multiple cores relative to both Koalas and DaskDF to efficiently execute pandas operations. It is notable that Koalas is often slower than pandas, due to the overhead of Spark. Details of the benchmark can be found in `our paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_.

Author: ```suggestion
**On operations supported by all systems, Modin provides substantial speedups.** Thanks to its optimized design, Modin is able to take advantage of multiple cores relative to both Koalas and DaskDF to efficiently execute pandas operations. It is notable that Koalas is often slower than pandas, due to the overhead of Spark. 
```

Author: Last sentence was repeated below.

Reviewer 2: LGTM

> . figure:: ../img/performance-not-all-supported.svg

Author: ```suggestion
.. figure:: ../img/performance-not-all-supported.svg
```

Reviewer 2: LGTM

Modin vs. Dask DataFrame vs. Koalas
===============================================

> Libraries such as `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF for short) and `Koalas <https://koalas.readthedocs.io/en/latest/>`_ aim to support the pandas API on top of distributed computing frameworks, Dask and Spark respectively. Instead, Modin aims to preserve the pandas API and semantics as is, while abstracting away the details of the distributed computing framework underneath. Thus, the aims of these libraries are fundamentally different.

Author: ```suggestion
Libraries such as `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`_ (DaskDF for short) and `Koalas <https://koalas.readthedocs.io/en/latest/>`_ aim to support the pandas API on top of distributed computing frameworks, Dask and Spark respectively. Instead, Modin aims to preserve the pandas API and behavior as is, while abstracting away the details of the distributed computing framework underneath. Thus, the aims of these libraries are fundamentally different.
```

**Modin's highly modular design is architected to run on a variety of systems, and support a variety of APIs.** The goal for the extensible design is that users can take the same notebook or script and seamlessly move between different clusters and environments, with Modin being able to support the pandas API on your preexisting infrastructure. Currently, Modin support running on Dask's compute engine in addition to Ray. The modular design makes it easier for developers to different execution engines or compile to different memory formats. Modin can run on a Dask cluster in the same way that DaskDF can, but they differ in the ways described above. In addition, Modin is continually expanding to support popular data processing APIs (SQL in addition to pandas, among other DSLs for data processing.) while leveraging the same underlying execution framework. Modin's flexible architecture also means that as the `pandas API continues to evolve <https://data-apis.org/blog/announcing_the_consortium/>`_, Modin can quickly move towards supporting new versions of the pandas API.

> .. figure:: ../img/performance-all-supported.svg

Author: These figures are really small and hard to read. Not sure if it's just showing up like this on my screen. Maybe we should lay them out on a 2X2 grid.

Author: Resolved in edited image

Ordering Semantics
------------------

By default, pandas preserves the order of the dataframe, so that users can expect a consistent, ordered view as they are operating on their dataframe. 

**Both DaskDF and Koalas make no guarantees about the order of rows in the DataFrame.**  This is because DaskDF sorts the ``index`` for optimization purposes to speed up computations that involve the row index; and as a result, it does not support user-specified order. Likewise, Koalas `does not support ordering <https://koalas.readthedocs.io/en/latest/whatsnew/v0.27.0.html#head-ordering>`_ by default because it will lead to a performance overhead when operating on distributed datasets. 

**DaskDF additionally does not support multi-indexing or sorting.** 
DaskDF sorts the data based on a single set of row labels for fast row lookups, and builds an indexing structure based on these row labels. Data is both logically and physically stored in the same order. As a result, DaskDF does not support a `sort` function.

> 

Reviewer 3: ```suggestion
```

**Modin reproduces the intuitive behavior in pandas where the order of the DataFrame is preserved, and supports multi-indexing.** Enforcing ordering on a parallel dataframe system like Modin requires non-trivial effort that involves decoupling of the logical and physical representation of the data, enabling the order to be lazily kept up-to-date, but eagerly computed based on user needs (See Section 4.2 in `our recent paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_). Modin abstracts away the physical representation of the data and provides an ordered view that is consistent with user's expectations.

> 

Reviewer 3: ```suggestion
```

Compatibility with Computational Frameworks

> -------------------------------------------
> **DaskDF and Koalas are meant to be run on Dask and Spark respectively.** They are highly tuned to the corresponding frameworks, and cannot be ported to other computational frameworks.

Reviewer 3: ```suggestion
-------------------------------------------

**DaskDF and Koalas are meant to be run on Dask and Spark respectively.** They are highly tuned to the corresponding frameworks, and cannot be ported to other computational frameworks.
```

> **Modin's highly modular design is architected to run on a variety of systems, and support a variety of APIs.** The goal for the extensible design is that users can take the same notebook or script and seamlessly move between different clusters and environments, with Modin being able to support the pandas API on your preexisting infrastructure. Currently, Modin support running on Dask's compute engine in addition to Ray. The modular design makes it easier for developers to different execution engines or compile to different memory formats. Modin can run on a Dask cluster in the same way that DaskDF can, but they differ in the ways described above. In addition, Modin is continually expanding to support popular data processing APIs (SQL in addition to pandas, among other DSLs for data processing.) while leveraging the same underlying execution framework. Modin's flexible architecture also means that as the `pandas API continues to evolve <https://data-apis.org/blog/announcing_the_consortium/>`_, Modin can quickly move towards supporting new versions of the pandas API.

Reviewer 3: ```suggestion
**Modin's highly modular design is architected to run on a variety of systems, and support a variety of APIs.** The goal for the extensible design is that users can take the same notebook or script and seamlessly move between different clusters and environments, with Modin being able to support the pandas API on your preexisting infrastructure. Currently, Modin support running on Dask's compute engine in addition to Ray. The modular design makes it easier for developers to different execution engines or compile to different memory formats. Modin can run on a Dask cluster in the same way that DaskDF can, but they differ in the ways described above. In addition, Modin is continually expanding to support popular data processing APIs (SQL in addition to pandas, among other DSLs for data processing) while leveraging the same underlying execution framework. Modin's flexible architecture also means that as the `pandas API continues to evolve <https://data-apis.org/blog/announcing_the_consortium/>`_, Modin can quickly move towards supporting new versions of the pandas API.
```

> 

Reviewer 3: ```suggestion
```

.. figure:: ../img/performance-all-supported.svg
   :align: center
   :alt: Scalability of operators supported by Modin and other systems

> 

Reviewer 3: ```suggestion
```

Performance Comparison

> ----------------------
> **On operations supported by all systems, Modin provides substantial speedups.** Thanks to its optimized design, Modin is able to take advantage of multiple cores relative to both Koalas and DaskDF to efficiently execute pandas operations. It is notable that Koalas is often slower than pandas, due to the overhead of Spark. 

Reviewer 3: ```suggestion
----------------------

**On operations supported by all systems, Modin provides substantial speedups.** Thanks to its optimized design, Modin is able to take advantage of multiple cores relative to both Koalas and DaskDF to efficiently execute pandas operations. It is notable that Koalas is often slower than pandas, due to the overhead of Spark. 
```

.. figure:: ../img/performance-not-all-supported.svg
   :align: center
   :alt: Scalability of operators supported by Modin but not by other systems

> **Modin provides substantial speedups even on operators not supported by other systems.** Thanks to its flexible partitioning schemes that enable it to support the vast majority of pandas operations — be it row, column, or cell-oriented - Modin provides benefits on operations such as ``joins``, ``median``, and ``infer_types``. While Koalas performs joins slower than Pandas, Dask failed to support joins on more than 20M rows, likely due poor support for `shuffles <https://coiled.io/blog/better-shuffling-in-dask-a-proof-of-concept/>`_. Details of the benchmark and additional join experiments can be found in `our paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_.

Reviewer 3: ```suggestion
**Modin provides substantial speedups even on operators not supported by other systems.** Thanks to its flexible partitioning schemes that enable it to support the vast majority of pandas operations — be it row, column, or cell-oriented - Modin provides benefits on operations such as ``join``(s), ``median``, and ``infer_types``. While Koalas performs ``join``(s) slower than Pandas, Dask failed to support ``join``(s) on more than 20M rows, likely due poor support for `shuffles <https://coiled.io/blog/better-shuffling-in-dask-a-proof-of-concept/>`_. Details of the benchmark and additional join experiments can be found in `our paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_.
```
?

Author: ```suggestion
**Modin provides substantial speedups even on operators not supported by other systems.** Thanks to its flexible partitioning schemes that enable it to support the vast majority of pandas operations — be it row, column, or cell-oriented - Modin provides benefits on operations such as ``join``, ``median``, and ``infer_types``. While Koalas performs ``join`` slower than Pandas, Dask failed to support ``join`` on more than 20M rows, likely due poor support for `shuffles <https://coiled.io/blog/better-shuffling-in-dask-a-proof-of-concept/>`_. Details of the benchmark and additional join experiments can be found in `our paper <https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf>`_.
```

> 
> 

Reviewer 3: ```suggestion
```
